js/views/chat/Chat.js

```diff
@@ -0,0 +1,457 @@
+    }
+  }
+
+  onSocketMessage(e) {
+    this.onNewChatMessage(e.jsonData.message);
+  }
+
+  onNewChatMessage(msg) {
+    if (msg && !msg.subject) {
+      const chatHead = this.collection.get(msg.peerId);
+      const chatHeadData = {
+        peerId: msg.peerId,
+        lastMessage: msg.message,
+        timestamp: msg.timestamp,
+        outgoing: false,
+        unread: 1,
+      };
+
+      if (this.conversation && this.conversation.guid === msg.peerId &&
+        this.conversation.isOpen) {
+        chatHeadData.unread = 0;
+      }
+
+      if (chatHead) {
+        if (!(this.conversation && this.conversation.guid === msg.peerId &&
+          this.conversation.isOpen)) {
+          chatHeadData.unread = chatHead.get('unread') + 1;
+        }
+
+        this.collection.remove(chatHead);
+      }
+
+      this.collection.add(chatHeadData, {
+        at: 0,
+        merge: true,
+      });
+    }
+  }
+
+  onConvoMarkedAsRead(guid) {
+    if (!guid) {
+      throw new Error('Please provide a guid.');
+    }
+
+    if (!this.conversation.isOpen) return;
+
+    const chatHead = this.collection.get(guid);
+
+    if (chatHead) chatHead.set('unread', 0);
+  }
+
+  open() {
+    if (this._isOpen) return;
+    this._isOpen = true;
+    getBody().addClass('chatOpen');
+  }
+
+  close() {
+    if (!this._isOpen) return;
+    this._isOpen = false;
+    getBody().removeClass('chatOpen');
+    this.closeConversation();
+  }
+
+  get isOpen() {
+    return this._isOpen;
+  }
+
+  /**
+   * This chat view, may need to know when it becomes visible,
+   * so please show it via this method.
+   */
+  show() {
+    this.$el.removeClass('hide');
+    return this;
+  }
+
+  hide() {
+    this.$el.addClass('hide');
+    return this;
+  }
+
+  /**
+   * This chat view mey need to know when it is attached to the dom,
+   * so please use this method to do so.
+   */
+  attach(container) {
+    if (!container || !(container instanceof $ && container[0] instanceof HTMLElement)) {
+      throw new Error('Please provide a container as a jQuery object or DOM element.');
+    }
+
+    $(container).append(this.el);
+    return this;
+  }
+
+  /**
+   * Adds css classes to our scroll element indicating whether the unread messages
+   * badges (top and / or bottom) need to be shown.
+   */
+  handleUnreadBadge() {
+    if (!this.chatHeads) return;
+
+    const firstUnreadChatHead = this.collection
+      .find(chatHead => (chatHead.get('unread')));
+
+    // todo: update isScrolledIntoView so that you could pass in an offset to
+    // determine if a certain portion of the el is out of view rather than the
+    // whole element
+
+    if (firstUnreadChatHead) {
+      const firstUnreadIndex = this.collection.indexOf(firstUnreadChatHead);
+
+      if (!isScrolledIntoView(this.chatHeads.views[firstUnreadIndex].el)) {
+        this.$el.addClass('outOfViewUnreadsAbove');
+      } else {
+        this.$el.removeClass('outOfViewUnreadsAbove');
```

Comment:
I need to know whether firstUnreadChatHead is true to determine even if I want to set the firstUnreadIndex variable. I could skip setting firstUnreadIndex and do:

```
this.$el.toggleClass('outOfViewUnreadsAbove, !!firstUnreadChatHead && !isScrolledIntoView(this.chatHeads.views[this.collection.indexOf(firstUnreadChatHead)].el));
```

But I think this is a case where a few extra lines is worth making things more readable.